### DISCUSSION TITLE
how to render shadow DOMs

### DISCUSSION THREAD

**Discussion by julisch94** [vuejs/test-utils#2250] [Q&A]
Hey folks, I'm using a component library that works with shadow DOM elements but when I call `render` in my test case and then print the DOM, I can't see any shadow DOM elements. How can I enable vue test utils to render shadow DOM elements, as well?

```ts
// navigation.spec.ts
import { describe, expect, it } from 'vitest'
import MyNavigationBar from './NavigationBar.vue'
import { mount } from '@vue/test-utils'

describe('NavigationBar', () => {
  it('displays a log out button', async () => {
    const wrapper = mount(MyNavigationBar)
    console.log(wrapper.html())
  })
})
```

Output in test:
```html
<ui-header-bar icon-label="Go to home page" icon-href="#">
  <ui-header-meta-link href="#">Logout</ui-header-meta-link>
</ui-header-bar>
```

Output in browser contains shadow DOMs with anchors and svg's.

**Comment by cexbrayat** [comment]
Hi @julisch94 

That's a good question.
`wrapper.html()` uses `stringifyNode`, as you can see here https://github.com/vuejs/test-utils/blob/d5f95cf0f628c3acf562f41807f9e92017261a23/src/baseWrapper.ts#L220-L235 which is defined here https://github.com/vuejs/test-utils/blob/d5f95cf0f628c3acf562f41807f9e92017261a23/src/utils/stringifyNode.ts#L1-L5 and uses `outerHTML£ 

So maybe outerHTML does not include the shadow DOM elements?

I think you can either:
- write your own 'html' function to serialize the wrapper, based on the one in VTU but which would serialize the shadow dom elements, but this might be a lot of work
- try to open a PR in VTU so that `wrapper.html()` handles this out of the box. If the changes are not too breaking I think we can land this.  Maybe it could be behind an option of `html()` (like the `raw` option)

**Reply by julisch94** [reply]
thanks for the quick reply! I'll see if I can find the time to put in the effort to aim for option 2. :)
